Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        async Task IDistributedCache.RefreshAsync(string key, CancellationToken token)
+        {
+            try
+            {
+                await _cacheEntries.Document(key).UpdateAsync(
+                    "LastRefresh", DateTime.UtcNow, cancellationToken: token);
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        void IDistributedCache.Remove(string key) =>
+            _cacheEntries.Document(key).DeleteAsync().Wait();
+
+        Task IDistributedCache.RemoveAsync(string key, CancellationToken token) =>
+            _cacheEntries.Document(key).DeleteAsync(cancellationToken: token);
+
+        /// <summary>
+        /// Creates a firestore document that stores the cache value and its
+        /// expiration info, but does not store it in firestore.
+        /// </summary>
+        /// <param name="value">The value to be stored in the cache.</param>
+        /// <param name="options">Expiration info.</param>
+        /// <returns></returns>
+        CacheDoc MakeCacheDoc(byte[] value, DistributedCacheEntryOptions options)
+        {
+            CacheDoc doc = new CacheDoc()
+            {
+                LastRefresh = DateTime.UtcNow,
+                Value = value,
+            };
+            if (options.SlidingExpiration.HasValue)
+            {
+                doc.SlidingExpirationSeconds =
+                    options.SlidingExpiration.Value.TotalSeconds;
+            }
+            else
+            {
+                var forever = DateTime.MaxValue - doc.LastRefresh.Value;
+                doc.SlidingExpirationSeconds = forever.TotalSeconds / 2;
+            }
+            if (options.AbsoluteExpiration.HasValue)
+            {
+                doc.AbsoluteExpiration = options.AbsoluteExpiration.Value.UtcDateTime;
+            }
+            if (options.AbsoluteExpirationRelativeToNow.HasValue)
+            {
+                doc.AbsoluteExpiration = DateTime.UtcNow +
+                    options.AbsoluteExpirationRelativeToNow.Value;
+            }
+            return doc;
+        }
+
+        void IDistributedCache.Set(string key, byte[] value, DistributedCacheEntryOptions options) =>
+            _cacheEntries.Document(key).SetAsync(MakeCacheDoc(value, options)).Wait();
+
+        Task IDistributedCache.SetAsync(string key, byte[] value, DistributedCacheEntryOptions options,

```

Comment:
I'd *definitely* unwrap the `Document.SetAsync` call here, and personally I'd unwrap the method declaration as well - that takes it to 132 characters, but I think it's still more readable that way.